jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -0,0 +1,465 @@
+ *     continues to produce.</dd>
+ *     <dt>ProduceInvokeConsume(PIC)</dt>
+ *     <dd>The producing thread consumes the task by invoking it with {@link Invocable#invokeNonBlocking(Runnable)}
+ *     and then continues to produce.</dd>
+ *     <dt>ProduceExecuteConsume(PEC)</dt>
+ *     <dd>The producing thread dispatches the task to a thread pool to be executed and then immediately resumes
+ *     producing.</dd>
+ *     <dt>ExecuteProduceConsume(EPC)</dt>
+ *     <dd>The producing thread consumes the task by executing it directly (as in PC mode) but then races with
+ *     a pending producer thread to take over production.
+ *     </dd>
+ * </dl>
+ * The sub-strategy is selected as follows:
+ * <dl>
+ *     <dt>PC</dt><dd>If the produced task has been invoked with {@link Invocable#invokeNonBlocking(Runnable)}
+ *     to indicate that it is {@link Invocable.InvocationType#NON_BLOCKING}.</dd>
+ *     <dt>EPC</dt><dd>If the producing thread is not {@link Invocable.InvocationType#NON_BLOCKING}
+ *     and a pending producer thread is available, either because there is already a pending producer
+ *     or one is successfully started with {@link TryExecutor#tryExecute(Runnable)}.</dd>
+ *     <dt>PIC</dt><dd>If the produced task has used the {@link Invocable#getInvocationType()} API to
+ *     indicate that it is {@link Invocable.InvocationType#EITHER}.</dd>
+ *     <dt>PEC</dt><dd>Otherwise.</dd>
+ * </dl>
+ *
+ * <p>Because of the preference for {@code PC} mode, on a multicore machine with many
+ * many {@link Invocable.InvocationType#NON_BLOCKING} tasks, multiple instances of the strategy may be
+ * required to keep all CPUs on the system busy.</p>
+ *
+ * <p>Since the producing thread may be invoked with {@link Invocable#invokeNonBlocking(Runnable)}
+ * this allows {@link AdaptiveExecutionStrategy}s to be efficiently and safely chained: so that a task
+ * produced by one execution strategy may become itself be a producer in a second execution strategy
+ * (e.g. an IO selector may use an execution strategy to handle multiple connections and each
+ * connection may use a execution strategy to handle multiplexed channels/streams within the connection).
+ * A task containing another {@link AdaptiveExecutionStrategy} should identify as
+ * {@link Invocable.InvocationType#EITHER} so when there are no pending producers threads available to
+ * the first strategy, then it may invoke the second as {@link Invocable.InvocationType#NON_BLOCKING}.
+ * This avoids starvation as the production on the second strategy can always be executed,
+ * but without the risk that it may block the last available producer for the first strategy.</p>
+ *
+ * <p>This strategy was previously named EatWhatYouKill (EWYK) because its preference to for
+ * a producer to directly consume a task was similar to the hunting proverb, in the sense that one
+ * should eat(consume) what they kill(produce).</p>
```

Comment:
```suggestion
 * <p>This strategy was previously named EatWhatYouKill (EWYK) because it prefers that
 * a producer directly consumes the task it produces, similarly to the hunting proverb that says that one
 * should eat (i.e. consume) what is killed (i.e. produced).</p>
```